File: backend/routers/analytics.py
from fastapi import APIRouter, Request
from pydantic import BaseModel
from typing import Literal
from db import supabase
from limiter import limiter

router = APIRouter(prefix="/analytics", tags=["analytics"])


class EventRequest(BaseModel):
    event_type: Literal["page_view", "demo_login", "register_company", "register_employee"]
    source: str = "main"


@router.post("/event")
@limiter.limit("60/minute")
async def track_event(request: Request, body: EventRequest):
    """Record an analytics event. Public endpoint — no auth required."""
    try:
        supabase.table("analytics_events").insert({
            "event_type": body.event_type,
            "source": body.source,
        }).execute()
    except Exception:
        pass  # fire-and-forget: never fail the caller
    return {"ok": True}

File: frontend/src/lib/analytics.ts
import { IS_DEMO } from "@/lib/demo";

const API = (import.meta.env.VITE_API_URL as string) || "http://localhost:8000";

export type AnalyticsEvent =
  | "page_view"
  | "demo_login"
  | "register_company"
  | "register_employee";

/**
 * Fire-and-forget analytics event. Uses plain fetch (not axios) so it never
 * triggers auth interceptors and never blocks or errors the caller.
 */
export function trackEvent(event: AnalyticsEvent): void {
  const source = IS_DEMO ? "demo" : "app";
  fetch(`${API}/analytics/event`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ event_type: event, source }),
  }).catch(() => {});
}

/** Track a page view once per browser session (survives SPA navigation). */
export function trackPageView(): void {
  const key = "__donkap_pv";
  if (!sessionStorage.getItem(key)) {
    sessionStorage.setItem(key, "1");
    trackEvent("page_view");
  }
}
